prov/shm/src/smr_msg.c

```diff
@@ -185,26 +185,45 @@ static ssize_t smr_generic_sendmsg(struct smr_ep *ep, const struct iovec *iov,
 
 	cmd = ofi_cirque_tail(smr_cmd_queue(peer_smr));
 
-	if (total_len > SMR_INJECT_SIZE || op_flags & FI_DELIVERY_COMPLETE) {
+	if (total_len <= SMR_MSG_DATA_LEN && !(op_flags & FI_DELIVERY_COMPLETE)) {
+		smr_format_inline(cmd, smr_peer_addr(ep->region)[peer_id].addr, iov,
+				  iov_count, op, tag, data, op_flags);
+	} else if (total_len <= SMR_INJECT_SIZE &&
+			!(op_flags & FI_DELIVERY_COMPLETE)) {
+		tx_buf = smr_freestack_pop(smr_inject_pool(peer_smr));
+		smr_format_inject(cmd, smr_peer_addr(ep->region)[peer_id].addr,
+				  iov, iov_count, op, tag, data, op_flags,
+				  peer_smr, tx_buf);
+	} else {
 		if (ofi_cirque_isfull(smr_resp_queue(ep->region))) {
 			ret = -FI_EAGAIN;
 			goto unlock_cq;
 		}
 		resp = ofi_cirque_tail(smr_resp_queue(ep->region));
 		pend = freestack_pop(ep->pend_fs);
-		smr_format_iov(cmd, smr_peer_addr(ep->region)[peer_id].addr, iov,
-			       iov_count, total_len, op, tag, data, op_flags,
-			       context, ep->region, resp, pend);
+		if (ep->region->cma_cap & SMR_CMA_CAP_ON) {
+			smr_format_iov(cmd, smr_peer_addr(ep->region)[peer_id].addr, iov,
+				       iov_count, total_len, op, tag, data, op_flags,
+				       context, ep->region, resp, pend);
+		} else {
+			/*
+			 * mmap protocol
+			 * Once SAR protocol gets merged, we need set a threshold
+			 * for switching from SAR to mmap
+			 */
```

Comment:
Can you remove the line mmap protocol line and turn this into a //TODO?

Reply:
Done.